dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/relationship/AbstractRelationshipService.java

```diff
@@ -363,20 +363,77 @@ else if ( daoRelationship.isDeleted() )
             return importSummary;
         }
 
-        org.hisp.dhis.relationship.Relationship _relationship = createDAORelationship( relationship );
-
-        daoRelationship.setRelationshipType( _relationship.getRelationshipType() );
-        daoRelationship.setTo( _relationship.getTo() );
-        daoRelationship.setFrom( _relationship.getFrom() );
-
-        relationshipService.updateRelationship( daoRelationship );
+        relationshipService.updateRelationship( updatedDAORelationship( daoRelationship, relationship ) );
 
         importSummary.setReference( daoRelationship.getUid() );
         importSummary.getImportCount().incrementUpdated();
 
         return importSummary;
     }
 
+    /**
+     * Update the relationship object fetched from the db only setting
+     * relationshipItem instance type. Using the same relationship and
+     * relationshipItem objects maintains hibernate session reference
+     *
+     * @param relationshipDb relationship fetched from db
+     * @param relationshipInput relationship in the payload
+     * @return relationshipDb with updated relationshipItems
+     */
+    private org.hisp.dhis.relationship.Relationship updatedDAORelationship(
+        org.hisp.dhis.relationship.Relationship relationshipDb, Relationship relationshipInput )
+    {
+        RelationshipType relationshipType = relationshipTypeCache.get( relationshipInput.getRelationshipType() );
+        relationshipDb.setRelationshipType( relationshipType );
+
+        RelationshipItem fromItem = relationshipDb.getFrom();
```

Comment:
Setting the from and to could probably be done in a separate method to reduce the code duplication :)